File: app/src/main/java/com/lumiyaviewer/lumiya/render/spatial/DrawListAvatarEntry.java
package com.lumiyaviewer.lumiya.render.spatial;

import com.lumiyaviewer.lumiya.render.avatar.DrawableAvatar;
import com.lumiyaviewer.lumiya.render.avatar.DrawableAvatarStub;
import com.lumiyaviewer.lumiya.slproto.objects.SLObjectAvatarInfo;
import java.lang.ref.WeakReference;
import javax.annotation.Nonnull;

public class DrawListAvatarEntry extends DrawListObjectEntry {
    private WeakReference<DrawableAvatar> drawableAvatar;
    private WeakReference<DrawableAvatarStub> drawableAvatarStub;

    @Nonnull
    private final SLObjectAvatarInfo objectAvatarInfo;

    public DrawListAvatarEntry(@Nonnull SLObjectAvatarInfo objectAvatarInfo) {
        super(objectAvatarInfo);
        this.drawableAvatar = null;
        this.drawableAvatarStub = null;
        this.objectAvatarInfo = objectAvatarInfo;
    }

    @Override
    public void addToDrawList(@Nonnull DrawList drawList) {
        if (drawList.avatars.size() >= drawList.avatarCountLimit && !this.objectAvatarInfo.isMyAvatar()) {
            WeakReference<DrawableAvatarStub> weakReference = this.drawableAvatarStub;
            DrawableAvatarStub drawableAvatarStub = weakReference != null ? weakReference.get() : null;
            if (drawableAvatarStub == null) {
                drawableAvatarStub = drawList.drawableStore.drawableAvatarStubCache.getUnchecked(this.objectAvatarInfo);
                this.drawableAvatarStub = new WeakReference<>(drawableAvatarStub);
            }
            drawList.avatarStubs.add(drawableAvatarStub);
            return;
        }
        WeakReference<DrawableAvatar> drawableAvatar2 = this.drawableAvatar;
        DrawableAvatar drawableAvatar = drawableAvatar2 != null ? drawableAvatar2.get() : null;
        if (drawableAvatar == null) {
            drawableAvatar = drawList.drawableStore.drawableAvatarCache.getUnchecked(this.objectAvatarInfo);
            this.drawableAvatar = new WeakReference<>(drawableAvatar);
        }
        drawList.avatars.add(drawableAvatar);
        if (this.objectAvatarInfo.isMyAvatar()) {
            drawList.myAvatar = drawableAvatar;
        }
    }

    @Nonnull
    public SLObjectAvatarInfo getObjectAvatarInfo() {
        return this.objectAvatarInfo;
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/render/spatial/DrawListAvatarEntry.kt
package com.lumiyaviewer.lumiya.render.spatial

import com.lumiyaviewer.lumiya.render.avatar.DrawableAvatar
import com.lumiyaviewer.lumiya.render.avatar.DrawableAvatarStub
import com.lumiyaviewer.lumiya.slproto.objects.SLObjectAvatarInfo
import java.lang.ref.WeakReference

class DrawListAvatarEntry(private val objectAvatarInfo: SLObjectAvatarInfo) : DrawListObjectEntry(objectAvatarInfo) {
    private var drawableAvatar: WeakReference<DrawableAvatar>? = null
    private var drawableAvatarStub: WeakReference<DrawableAvatarStub>? = null

    override fun addToDrawList(drawList: DrawList) {
        if (drawList.avatars.size >= drawList.avatarCountLimit && !objectAvatarInfo.isMyAvatar) {
            var drawableAvatarStub = this.drawableAvatarStub?.get()
            if (drawableAvatarStub == null) {
                drawableAvatarStub = drawList.drawableStore.drawableAvatarStubCache.getUnchecked(objectAvatarInfo)!!
                this.drawableAvatarStub = WeakReference(drawableAvatarStub)
            }
            drawList.avatarStubs.add(drawableAvatarStub)
            return
        }
        var drawableAvatar = this.drawableAvatar?.get()
        if (drawableAvatar == null) {
            drawableAvatar = drawList.drawableStore.drawableAvatarCache.getUnchecked(objectAvatarInfo)!!
            this.drawableAvatar = WeakReference(drawableAvatar)
        }
        drawList.avatars.add(drawableAvatar)
        if (objectAvatarInfo.isMyAvatar) {
            drawList.myAvatar = drawableAvatar
        }
    }

    fun getObjectAvatarInfo(): SLObjectAvatarInfo = objectAvatarInfo
}
